corgi/core/models.py

```diff
@@ -902,143 +902,125 @@ def get_product_details(variant_names: tuple[str], stream_names: list[str]) -> d
-    ) -> "ComponentQuerySet":
-        """Return an ofuri's components from latest builds across all product streams."""
-
-        # the concept of 'latest component' is only relevent within product boundaries
-        # we want to constrain by product type/ofuri which is why we pass in model_type and ofuri
-        # into the get_latest_component stored proc annotation
-
-        components = self
-        if not (include_non_root_components):
-            components = (
-                components.select_related("software_build")
-                .prefetch_related("productstreams")
-                .root_components()
-                .using("read_only")
-            )
-        if ofuri:
-            components = components.filter(productstreams__ofuri=ofuri).distinct()
-
-        latest_components_uuids = set(
-            components.values("name", "namespace")
-            .order_by("name")  # required to avoid cross-join fun
-            .distinct("name", "namespace")
+    @staticmethod
+    def _latest_components_func(
+        components: "ComponentQuerySet", model_type: str, ofuri: str, include_inactive_streams: bool
+    ) -> Iterable[str]:
+        return (
+            components.values("type", "namespace", "name", "arch")
+            .order_by()  # required to avoid cross-join fun
+            .distinct()
             .annotate(
                 latest_version=Func(
                     Value(model_type),
                     Value(ofuri),
+                    F("type"),
                     F("namespace"),
                     F("name"),
+                    F("arch"),
                     Value(include_inactive_streams),
                     function="get_latest_component",
                     output_field=models.UUIDField(),
                 )
             )
             .values_list("latest_version", flat=True)
-            .using("read_only")
+        )
+
+    def latest_components(
+        self,
+        ofuri: str,
+        model_type: str = "ProductStream",
+        include: bool = True,
+        include_inactive_streams=False,
+    ) -> "ComponentQuerySet":
+        """Return components from latest builds in a single stream."""
+
+        # the concept of 'latest component' is only relevant within product boundaries
+        # we want to constrain by product type/ofuri which is why we pass in model_type and ofuri
+        # into the get_latest_component stored proc annotation
+
+        product_prefetch = "productstreams"
+        if model_type == "Product":
+            product_prefetch = "products"
+        if model_type == "ProductVersion":
+            product_prefetch = "productversions"
+        if model_type == "ProductVariant":
+            product_prefetch = "productvariants"
+        components = (
+            self.root_components()
+            .select_related("software_build")
+            .prefetch_related(product_prefetch)
+            .filter(productstreams__ofuri=ofuri)
+        )
```

Comment:
We always filter on productstreams__ofuri here, even if the user is looking at a different model.
```suggestion
        lookup = {f"{product_prefetch}__ofuri": ofuri}
        components = (
            self.root_components()
            .select_related("software_build")
            .prefetch_related(product_prefetch)
            .filter(**lookup)
        )